File: 1396-count-servers-that-communicate/count-servers-that-communicate.py
# declare two lookup tables that will count the number of servers in each row and columns
# iterate throught the grid
# if the current server has more than one server in its row or column, increment the count of servers
# return the count of servers

class Solution:
    def countServers(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        columns = len(grid[0])
        row_count = {}
        column_count = {}
        servers = 0

        for i in range(rows):
            for j in range(columns):
                cell = grid[i][j]

                if (cell == 1):
                    if (i not in row_count):
                        row_count[i] = 0
                    row_count[i] +=1

                    if (j not in column_count):
                        column_count[j] = 0
                    column_count[j] += 1

        for i in range(rows):
            for j in range(columns):
                cell = grid[i][j]

                if (cell == 1):
                    if (row_count[i] > 1 or column_count[j] > 1):
                        servers += 1

        return servers
                
        
        

File: 1396-count-servers-that-communicate/count-servers-that-communicate.js
// take the dimensions of the grid
// declare a variable lookup that will store servers in rows and columns
// iterate through the grid
// for each server 
// if there is a server present on that cell or on that row in the lookup object increment the server count by 1
// else add the cell to the row and column in the lookup  object

var countServers = function(grid) {
    let rows = grid.length
    let columns = grid[0].length
    let rowCount = {}
    let columnCount = {}
    let serverCount = 0

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < columns; j++) {
            cell = grid[i][j]
            if (cell == 1) {
                if (!rowCount[i]) rowCount[i] = 0
                rowCount[i] ++
                if (!columnCount[j]) columnCount[j] = 0
                columnCount[j] ++
            }
        }
    }

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < columns; j++) {
            cell = grid[i][j]
            if (cell == 1) {
                if (rowCount[i] > 1 || columnCount[j] > 1) serverCount ++
            }
        }
    }

    return serverCount
};